modules/substrate/src/verifier.rs

```diff
@@ -163,12 +185,16 @@ where
 				let height = (*header.number())
 					.checked_add(&change.delay)
 					.ok_or(ImportError::ScheduledHeightOverflow)?;
+
 				let scheduled_change = ScheduledChange {
 					authority_set: next_set,
 					height,
 				};
 
-				self.storage.schedule_next_set_change(scheduled_change);
+				// Note: It's important that the signal hash is updated if a header schedules a
+				// change or else we end up with inconsistencies in other places.
+				signal_hash = Some(hash);
+				self.storage.schedule_next_set_change(hash, scheduled_change);
```

Comment:
IMHO storage writes should be bundled together at the very end of that function. It's too easy to add another condition right after this line with `?` at the end which might have catastrophic consequences.

Reply:
I agree, but it's a little annoying to change at the moment. I'll make a note of it though